File: packages/extension/src/conversationClient.ts
import { defaultSocketFactory, type ClientSocket, type SocketFactory } from "./socket";

export type { ClientSocket };

export interface ConversationClientOptions {
  /** `ws://127.0.0.1:<conversationPort>` (from discovery). */
  url: string;
  /** Per-session token presented in the first frame. */
  token: string;
  /** Called for each panel beat streamed by the daemon. Opaque here (the panel applies it). */
  onBeat: (beat: unknown) => void;
  /** Injected for tests; defaults to a real `WebSocket` adapter. */
  socketFactory?: SocketFactory;
  onClose?: () => void;
}

export interface ConversationClient {
  /** Begin a task (queued until the socket is open + authenticated). */
  start(task: string): void;
  /** Stop the in-flight turn. */
  stop(): void;
  /** Close the connection. */
  close(): void;
}

/**
 * Connect the side panel to the daemon's conversation WS: on open it sends the
 * `{ type: "auth", token }` handshake first, then streams `{ type: "beat", beat }`
 * frames to `onBeat` and sends `start`/`stop` commands. Sends issued before the
 * socket opens are queued and flushed (after auth) once it does.
 */
export function connectConversation(opts: ConversationClientOptions): ConversationClient {
  const socket = (opts.socketFactory ?? defaultSocketFactory)(opts.url);
  let open = false;
  let closed = false;
  const queue: string[] = [];

  const send = (value: unknown): void => {
    if (closed) return; // never send/queue on a closed connection
    const frame = JSON.stringify(value);
    if (open) socket.send(frame);
    else queue.push(frame);
  };

  socket.onOpen(() => {
    open = true;
    // Auth is fire-and-forget (no `auth_ok` wait — the daemon closes on bad auth);
    // it must be the first frame, before any queued start/stop.
    socket.send(JSON.stringify({ type: "auth", token: opts.token }));
    for (const frame of queue) socket.send(frame);
    queue.length = 0;
  });

  socket.onMessage((data) => {
    if (closed) return;
    let msg: unknown;
    try {
      msg = JSON.parse(data);
    } catch {
      return; // ignore malformed frames
    }
    if (typeof msg === "object" && msg !== null && (msg as { type?: unknown }).type === "beat") {
      opts.onBeat((msg as { beat: unknown }).beat);
    }
  });

  socket.onClose(() => {
    closed = true;
    open = false;
    opts.onClose?.();
  });

  return {
    start: (task) => send({ type: "start", task }),
    stop: () => send({ type: "stop" }),
    close: () => socket.close(),
  };
}

File: packages/mac-shell/Sources/FairyShell/ConversationClient.swift
import Foundation

/// Drives a daemon conversation over an injected `ConversationSocket`: sends the
/// `{type:auth,token}` handshake first on open, flushes any queued commands, decodes
/// inbound `{type:"beat",beat}` frames to `onBeat` (the beat passed through as raw
/// JSON for the WebView), and encodes `start`/`stop`/`resolveProposal` commands
/// (queued until open). Mirrors the extension's `connectConversation` semantics.
public final class ConversationClient {
  private let socket: ConversationSocket
  private let token: String
  private let onBeat: (String) -> Void
  private var open = false
  private var closed = false
  private var queue: [String] = []

  public init(socket: ConversationSocket, token: String, onBeat: @escaping (String) -> Void) {
    self.socket = socket
    self.token = token
    self.onBeat = onBeat
    socket.onOpen { [weak self] in self?.handleOpen() }
    socket.onText { [weak self] in self?.handleText($0) }
    socket.onClose { [weak self] in self?.handleClose() }
  }

  /// Open the socket (the handshake is sent automatically once it opens).
  public func connect() { socket.connect() }

  public func start(_ task: String) { send(["type": "start", "task": task]) }
  public func stop() { send(["type": "stop"]) }

  /// Resolve a save proposal. `proposalJSON` is the opaque proposal object as JSON
  /// (the panel produced it); it's embedded verbatim into the frame.
  public func resolveProposal(_ proposalJSON: String) {
    guard let data = proposalJSON.data(using: .utf8),
          let proposal = try? JSONSerialization.jsonObject(with: data) else { return }
    send(["type": "resolveProposal", "proposal": proposal, "accept": true])
  }

  public func close() {
    closed = true
    open = false
    queue.removeAll()       // drop anything queued before open so a late open can't flush it
    socket.close()
  }

  // MARK: - Socket callbacks

  private func handleOpen() {
    guard !closed else { return }  // closed before the socket opened — send nothing
    open = true
    socket.send(encode(["type": "auth", "token": token]))  // auth must be the first frame
    for frame in queue { socket.send(frame) }
    queue.removeAll()
  }

  private func handleText(_ text: String) {
    guard !closed,
          let obj = (try? JSONSerialization.jsonObject(with: Data(text.utf8))) as? [String: Any],
          (obj["type"] as? String) == "beat",
          let beat = obj["beat"],
          let beatData = try? JSONSerialization.data(withJSONObject: beat),
          let beatJSON = String(data: beatData, encoding: .utf8) else { return }
    onBeat(beatJSON)
  }

  private func handleClose() {
    open = false
    closed = true
  }

  // MARK: - Outbound

  private func send(_ value: [String: Any]) {
    if closed { return }
    let frame = encode(value)
    if open { socket.send(frame) } else { queue.append(frame) }
  }

  private func encode(_ value: [String: Any]) -> String {
    guard let data = try? JSONSerialization.data(withJSONObject: value),
          let s = String(data: data, encoding: .utf8) else { return "{}" }
    return s
  }
}
